File: frontend/src/api/resume.ts
import { fetchApi } from '../lib/api';

export interface PersonalInfo {
  fullName: string;
  email: string;
  phone: string;
  location?: string;
  linkedin?: string;
  github?: string;
  portfolio?: string;
  careerObjective?: string;
}

export interface EducationItem {
  institution: string;
  degree: string;
  fieldOfStudy: string;
  startDate: string;
  endDate: string;
  grade?: string;
}

export interface SkillCategory {
  category: string;
  skills: string[];
}

export interface ExperienceItem {
  company: string;
  role: string;
  location?: string;
  startDate: string;
  endDate: string;
  description: string[];
}

export interface ProjectItem {
  title: string;
  technologies: string[];
  link?: string;
  description: string[];
}

export interface CertificationItem {
  name: string;
  issuer: string;
  issueDate?: string;
}

export interface AchievementItem {
  title: string;
  description?: string;
}

export interface ReferenceItem {
  name: string;
  title: string;
  company: string;
  email?: string;
  phone?: string;
}

export interface ResumePayload {
  title: string;
  template: 'modern' | 'professional' | 'minimal' | 'corporate' | 'student';
  personalInfo: PersonalInfo;
  education: EducationItem[];
  skills: SkillCategory[];
  experience: ExperienceItem[];
  projects: ProjectItem[];
  certifications: CertificationItem[];
  achievements: AchievementItem[];
  languages: string[];
  hobbies: string[];
  references: ReferenceItem[];
}

export interface ResumeResponse {
  id: number;
  user_id: number;
  title: string;
  template: string;
  resume_data: ResumePayload;
  ats_score: number;
  grammar_score: number;
  formatting_score: number;
  missing_skills: string[];
  suggestions: string[];
  updated_at?: string;
}

export interface ATSResult {
  overall_score: number;
  grammar_score: number;
  formatting_score: number;
  missing_skills: string[];
  weak_sections: string[];
  suggestions: string[];
}

export const getResume = async (): Promise<ResumeResponse> => {
  return await fetchApi('/resume');
};

export const saveResume = async (payload: ResumePayload): Promise<ResumeResponse> => {
  return await fetchApi('/resume/create', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
};

export const deleteResume = async (): Promise<void> => {
  await fetchApi('/resume', { method: 'DELETE' });
};

export const enhanceResumeContent = async (
  section: string,
  content: string,
  context?: string
): Promise<string> => {
  const res = await fetchApi('/resume/enhance', {
    method: 'POST',
    body: JSON.stringify({ section, content, context }),
  });
  return res.enhanced_text;
};

export const evaluateATS = async (
  resume_data: ResumePayload,
  job_description?: string
): Promise<ATSResult> => {
  return await fetchApi('/resume/ats', {
    method: 'POST',
    body: JSON.stringify({ resume_data, job_description }),
  });
};

File: backend/app/schemas/resume.py
from pydantic import BaseModel, EmailStr, HttpUrl
from typing import Optional, List, Dict, Any
from datetime import datetime


class PersonalInfo(BaseModel):
    fullName: str = "Rahul Verma"
    email: str = "rahul.verma@college.edu"
    phone: str = "+91 98765 43210"
    location: Optional[str] = "Chennai, India"
    linkedin: Optional[str] = "https://linkedin.com/in/rahulverma"
    github: Optional[str] = "https://github.com/rahulverma"
    portfolio: Optional[str] = "https://rahulverma.dev"
    careerObjective: Optional[str] = "Motivated Computer Science student looking for Software Engineering roles."


class EducationItem(BaseModel):
    institution: str
    degree: str
    fieldOfStudy: str
    startDate: str
    endDate: str
    grade: Optional[str] = ""


class SkillCategory(BaseModel):
    category: str  # e.g., "Languages", "Frameworks", "Tools"
    skills: List[str]


class ExperienceItem(BaseModel):
    company: str
    role: str
    location: Optional[str] = ""
    startDate: str
    endDate: str
    description: List[str]


class ProjectItem(BaseModel):
    title: str
    technologies: List[str]
    link: Optional[str] = ""
    description: List[str]


class CertificationItem(BaseModel):
    name: str
    issuer: str
    issueDate: Optional[str] = ""


class AchievementItem(BaseModel):
    title: str
    description: Optional[str] = ""


class ReferenceItem(BaseModel):
    name: str
    title: str
    company: str
    email: Optional[str] = ""
    phone: Optional[str] = ""


class ResumePayload(BaseModel):
    title: str = "My Primary Resume"
    template: str = "modern"  # modern, professional, minimal, corporate, student
    personalInfo: PersonalInfo
    education: List[EducationItem] = []
    skills: List[SkillCategory] = []
    experience: List[ExperienceItem] = []
    projects: List[ProjectItem] = []
    certifications: List[CertificationItem] = []
    achievements: List[AchievementItem] = []
    languages: List[str] = []
    hobbies: List[str] = []
    references: List[ReferenceItem] = []


class ResumeResponse(BaseModel):
    id: int
    user_id: int
    title: str
    template: str
    resume_data: ResumePayload
    ats_score: int
    grammar_score: int
    formatting_score: int
    missing_skills: List[str] = []
    suggestions: List[str] = []
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ATSEvaluationRequest(BaseModel):
    job_description: Optional[str] = ""
    resume_data: ResumePayload


class AIEnhanceRequest(BaseModel):
    section: str  # "careerObjective" | "experience" | "project" | "skills"
    content: str
    context: Optional[str] = ""
